File: Easy/20-ValidParentheses.cc
#include <iostream>
#include <stack>
#include <unordered_map>

using namespace std;


// Description:  https://leetcode.com/problems/valid-parentheses/


// Approach: Stack

// When encountering a closing bracket, check if it matches the top of the stack.

/* 
    Time complexity: O(n) - We traverse the string once.
    Space complexity: O(n) - In the worst case, we store all opening brackets in the stack.


    Use a stack to keep track of opening brackets. 
    When encountering a closing bracket, check if it matches the top of the stack.
    If it matches, pop the top of the stack. If it doesn't match or the stack is empty, the string is invalid.
    After processing all characters, if the stack is empty, the string is valid; otherwise, it is invalid.
*/

class Solution {
public:
    bool isValid(string s) {
        stack<char> openedBrackets;
        unordered_map<char, char> bracketMap = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };

        for (char c : s) {
            if (bracketMap.find(c) != bracketMap.end()) { // closed bracket case
                if (openedBrackets.empty()|| openedBrackets.top() != bracketMap[c]) return false;
                openedBrackets.pop();
            }
            else  openedBrackets.push(c);
        }
        
        return openedBrackets.empty();
    }
};


//Test
int main() {
    Solution solution;
    string test1 = "()";
    string test2 = "()[]{}";
    string test3 = "(]";

    cout << "Test case 1: " << (solution.isValid(test1) ? "true" : "false") << endl; // Output: true
    cout << "Test case 2: " << (solution.isValid(test2) ? "true" : "false") << endl; // Output: true
    cout << "Test case 3: " << (solution.isValid(test3) ? "true" : "false") << endl; // Output: false


}

File: Python/Easy/20-ValidParentheses.py
# Description: https://leetcode.com/problems/valid-parentheses/description/

"""
    Approach: Stack with Dictionary

    Use a list as a stack to track opening brackets during iteration.
    Store bracket pairs in a dictionary for O(1) matching lookups.
    For each opening bracket, append it to the stack.
    For each closing bracket, check if stack is non-empty and top matches.
    If mismatch or stack empty on closing bracket, return False.
    After processing, stack should be empty for valid parentheses.

    Time complexity: O(n) - Single pass through the string.
    Space complexity: O(n) - Stack stores at most n/2 opening brackets.
"""

from typing import Dict

# Solution
class Solution:
    def isValid(self, s: str) -> bool:
        openedBrackets = []
        bracketMap = {
            ')': '(',
            ']': '[',
            '}': '{'
        }
        
        for c in s:
            if c in bracketMap:  # closed bracket case
                if not openedBrackets or openedBrackets[-1] != bracketMap[c]:
                    return False
                openedBrackets.pop()
            else:
                openedBrackets.append(c)
        
        return len(openedBrackets) == 0

# Test
if __name__ == "__main__":
    solution = Solution()
    test1 = "()"
    test2 = "()[]{}"
    test3 = "(]"
    
    print(f"Input: {test1}")
    print(f"Output: {solution.isValid(test1)}")  # Output: True
    
    print(f"Input: {test2}")
    print(f"Output: {solution.isValid(test2)}")  # Output: True
    
    print(f"Input: {test3}")
    print(f"Output: {solution.isValid(test3)}")  # Output: False
